xonsh/aliases.py

```diff
@@ -436,13 +531,50 @@ def partial_eval_alias(f, acc_args=()):
             numargs += 1
         elif name in ALIAS_KWARG_NAMES and param.kind == param.KEYWORD_ONLY:
             numargs += 1
-    if numargs < 7:
+    if numargs < 8:
         return PARTIAL_EVAL_ALIASES[numargs](f, acc_args=acc_args)
     else:
-        e = "Expected proxy with 6 or fewer arguments for {}, not {}"
+        e = "Expected proxy with 7 or fewer arguments for {}, not {}"
         raise XonshError(e.format(", ".join(ALIAS_KWARG_NAMES), numargs))
 
 
+def run_alias_by_params(func: tp.Callable, params: dict[str, tp.Any]):
+    """
+    Run alias function based on signature and params.
+    If function param names are in alias signature fill them.
+    If function params have unknown names fill using alias signature order.
+    """
+    alias_params = OrderedDict(
+        {
+            "args": None,
+            "stdin": None,
+            "stdout": None,
+            "stderr": None,
+            "spec": None,
+            "stack": None,
+            "spec_modifiers": None,
+        }
+    )
+    alias_params |= params
+    sign = inspect.signature(func)
+    func_params = sign.parameters.items()
+    kwargs = {
+        name: alias_params[name] for name, p in func_params if name in alias_params
+    }
+
+    if len(kwargs) != len(func_params):
+        # There is unknown param. Switch to positional mode.
+        kwargs = OrderedDict()
+        vals = list(alias_params.values())
+        len_vals = len(vals)
+        i = 0
+        for namep in func_params:
+            kwargs[namep[0]] = vals[i]
+            if (i := i + 1) == len_vals:
+                break
```

Comment:
I'm pretty sure this behavior is expressable as a single expression

```suggestion
        kwargs = {
            namep[0]: val
            for namep, val in zip(func_params, alias_params.values())
        }
```

or

```suggestion
        kwargs = dict(zip(map(operator.itemgetter(0), func_params), alias_params.values()))
```

And of course, you can use `OrderedDict` if you want, but with Python 3.6+, the std dict will suffice.